core/space/services/services_fs.go

```diff
@@ -156,29 +156,125 @@ func (s *Space) createFolder(ctx context.Context, path string, b textile.Bucket)
 	return root.String(), nil
 }
 
-func (s *Space) AddItems(ctx context.Context, sourcePaths []string, targetPath string) (<-chan domain.AddItemResult, error) {
+func (s *Space) AddItems(ctx context.Context, sourcePaths []string, targetPath string) (<-chan domain.AddItemResult, domain.AddItemsResponse, error) {
+	// check if all sourcePaths exist, else return err
+	for _, sourcePath := range sourcePaths {
+		if !PathExists(sourcePath) {
+			return nil, domain.AddItemsResponse{}, errors.New(fmt.Sprintf("path not found at %s", sourcePath))
+		}
+	}
 	// TODO: add support for bucket slug
 	b, err := s.tc.GetDefaultBucket(ctx)
 	if err != nil {
-		return nil, err
+		return nil, domain.AddItemsResponse{}, err
 	}
 	results := make(chan domain.AddItemResult)
+
+	totalsRes, err := getTotals(RemoveDuplicates(sourcePaths))
+	if err != nil {
+		return nil, domain.AddItemsResponse{}, err
+	}
 	go func() {
 		s.addItems(ctx, RemoveDuplicates(sourcePaths), targetPath, b, results)
 		close(results)
 	}()
 
-	return results, nil
+	return results, totalsRes, nil
 }
 
-func (s *Space) addItems(ctx context.Context, sourcePaths []string, targetPath string, b textile.Bucket, results chan<- domain.AddItemResult) error {
-	// check if all sourcePaths exist, else return err
+// get totals for addItems operation
+func getTotals(sourcePaths []string) (domain.AddItemsResponse, error) {
+	var wg sync.WaitGroup
+	wg.Add(len(sourcePaths))
+	filesRes := make(chan domain.AddItemsResponse)
+	results := make([]domain.AddItemsResponse, 0)
 	for _, sourcePath := range sourcePaths {
-		if !PathExists(sourcePath) {
-			return errors.New(fmt.Sprintf("path not found at %s", sourcePath))
+		go func(pathInFs string) {
+			defer wg.Done()
+			if IsPathDir(pathInFs) {
+				// counting folder as a file in total with 0 bytes
+				filesRes <- domain.AddItemsResponse{
+					TotalFiles: 1,
+					TotalBytes: 0,
+				}
+				// get recursive
+				var folderSubPaths []string
+				files, err := ioutil.ReadDir(pathInFs)
+				if err != nil {
+					log.Error(fmt.Sprintf("error reading folder path %s ", pathInFs), err)
+					filesRes <- domain.AddItemsResponse{
+						Error: err,
+					}
+					return
+				}
+				for _, file := range files {
+					subPath := pathInFs + "/" + file.Name()
+					if subPath != pathInFs {
```

Comment:
Was having a little trouble understanding this piece.  This means when the contents are a file `subPath == pathInFs`?

Reply:
yeah, not sure its needed honestly, I think i had an issue when using `Walk` method that needed that check, this is a precaution to avoid adding the existing path to the subpaths and dont get into an infinite recursive loop.